satellite/overlay/selectednodescache.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package overlay
+
+import (
+	"context"
+	"math/rand"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	"storj.io/common/storj"
+)
+
+// SelectedNodesCache keeps a list of all the storage nodes that are qualified to store data
+// We organize the nodes by if they are reputable or a new node on the network.
+// The cache will get refreshed once the staleness time has past.
+type SelectedNodesCache struct {
+	log              *zap.Logger
+	db               DB
+	nodeSelectionCfg NodeSelectionConfig
+	staleness        time.Duration
+	lastRefresh      time.Time
+
+	reputableMu    sync.Mutex
+	reputableNodes []CachedNode
+
+	newNodesMu sync.Mutex
+	newNodes   []CachedNode
```

Comment:
Make a state struct similarly how reliability cache has. That way it's harder to make bugs related with update boundaries.